src/contracts/fund/trading/exchanges/MelonEngine.ts

```diff
@@ -0,0 +1,75 @@
+import BigNumber from 'bignumber.js';
+import { Address } from '../../../../Address';
+import { sameAddress } from '../../../../utils/sameAddress';
+import { Trading } from '../Trading';
+import { encodeFunctionSignature } from '../../../../utils/encodeFunctionSignature';
+import { ExchangeAdapterAbi } from '../../../../abis/ExchangeAdapter.abi';
+import { zeroAddress } from '../../../../utils/zeroAddress';
+import { MelonEngineNotRegisteredWithFundError } from '../Trading.errors';
+import { checkSufficientBalance } from '../utils/checkSufficientBalance';
+import { checkFundIsNotShutdown } from '../utils/checkFundIsNotShutdown';
+
+export interface TakeOrderMelonEngine {
+  makerAsset: Address;
+  takerAsset: Address;
+  makerQuantity: BigNumber;
+  takerQuantity: BigNumber;
+}
+
+export class MelonEngine {
+  constructor(public readonly trading: Trading, public readonly exchangeIndex: number) {}
+
+  public static async create(trading: Trading, address: Address) {
+    const exchangeIndex = await this.index(trading, address);
+    if (exchangeIndex === null) {
+      throw new MelonEngineNotRegisteredWithFundError(address);
+    }
+
+    return new this(trading, exchangeIndex);
+  }
+
+  public static async index(trading: Trading, address: Address) {
```

Comment:
This is used in both Kyber and MelonEngine. Maybe create a Base class that both extend from?